crates/iota-mainnet-unlocks/src/unlocks.rs

```diff
@@ -0,0 +1,254 @@
+#[derive(Debug, Clone, Deserialize)]
+pub struct StillLockedEntry {
+    /// UTC timestamp at which the tokens are still locked.
+    pub timestamp: DateTime<Utc>,
+    /// Total locked amount (nano-units) still locked at the timestamp.
+    pub amount_still_locked: u64,
+}
+
+/// In-memory store holding the aggregated token unlock data.
+#[derive(Debug, Clone)]
+pub struct AggregatedUnlocksStore {
+    // Each entry represents the total number of tokens still locked at the specific point in time.
+    entries: BTreeMap<DateTime<Utc>, StillLockedEntry>,
+}
+
+impl AggregatedUnlocksStore {
+    /// Creates a new `AggregatedUnlocksStore` based on the provided `Chain`.
+    ///
+    /// For `Chain::Mainnet`, this method loads the precompiled unlock data from
+    /// file, since the unlock data is only valid for mainnet. For any other
+    /// chain (e.g. testnet, devnet), it returns an empty store, as no
+    /// precompiled unlock data is applicable.
+    pub fn from_chain(chain: Chain) -> Result<Self> {
+        match chain {
+            Chain::Mainnet => Self::load_mainnet_unlocks(),
+            // The precompiled unlock data is only valid for mainnet. For other chains, return an
+            // empty store.
+            _ => Ok(Self::default()),
+        }
+    }
+
+    /// Loads the aggregated token unlock data from the given JSON file at the
+    /// crate root.
+    fn load_mainnet_unlocks() -> Result<Self> {
+        let crate_dir = env!("CARGO_MANIFEST_DIR");
+        let path = PathBuf::from(crate_dir)
+            .join("data")
+            .join(AGGREGATED_MAINNET_UNLOCKS_FILE);
+
+        let data = fs::read_to_string(&path)
+            .with_context(|| format!("could not read locked supply file: {:?}", path))?;
+
+        Self::from_json(&data)
+    }
+
+    /// Parses the given JSON string into a `TokenUnlocksStore`.
+    fn from_json(json: &str) -> Result<Self> {
+        let parsed: Vec<StillLockedEntry> =
+            serde_json::from_str(json).context("invalid JSON format in unlock data")?;
+
+        let mut map = BTreeMap::new();
+        for entry in parsed {
+            if map.contains_key(&entry.timestamp) {
+                return Err(anyhow::anyhow!(
+                    "duplicate entry found for timestamp: {}",
+                    entry.timestamp
+                ));
+            }
+            map.insert(entry.timestamp, entry);
+        }
+
+        Ok(Self { entries: map })
+    }
+
+    /// Returns the total amount of tokens (in nano-units) that are still locked
+    /// at the given timestamp.
+    pub fn still_locked_tokens(&self, timestamp: DateTime<Utc>) -> u64 {
+        self.entries
+            .range(..=timestamp)
+            .next_back()
+            .map(|(_, entry)| entry.amount_still_locked)
+            .unwrap_or_else(|| {
+                // No earlier entries exist: use first available as retroactively valid
+                self.entries
+                    .iter()
+                    .next()
```

Comment:
You can use [first_entry](https://doc.rust-lang.org/std/collections/struct.BTreeMap.html#method.first_entry)

Reply:
It requires the map to be mutable. I think we should better keep it like it is